modules/nf-core/rgi/main/main.nf

```diff
@@ -24,7 +27,53 @@ process RGI_MAIN {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+
+    def db_download_cmd = ""
+    def db_card_cmd = ""
+    def db_wildcard_cmd = ""
+    def load_wildcard = ""
+    def copy_db_cmd = "cp $database/card* database_output"
+
+    if (wildcard.equalsIgnoreCase("yes")) {
+        if (!database) { // Download both DBs if no path is given
+            db_download_cmd = "wget https://card.mcmaster.ca/latest/data;tar -xvf data ./card.json -C database_output;wget -O wildcard_data.tar.bz2 https://card.mcmaster.ca/latest/variants;mkdir -p wildcard;tar -xjf wildcard_data.tar.bz2 -C wildcard;gunzip wildcard/*.gz"
+            db_card_cmd = "rgi card_annotation -i database_output/card.json > card_annotation.log 2>&"
+            db_wildcard_cmd = "rgi wildcard_annotation -i wildcard --card_json database_output/card.json -v \$DB_VERSION > wildcard_annotation.log 2>&1"
+            database = "."
+        }
+        load_wildcard = """ \\
+            --wildcard_annotation database_output/wildcard_database_v\$DB_VERSION.fasta \\
+            --wildcard_annotation_all_models database_output/wildcard_database_v\$DB_VERSION\\_all.fasta \\
+            --wildcard_index database_output/wildcard/index-for-model-sequences.txt \\
+            --amr_kmers database_output/wildcard/all_amr_61mers.txt \\
+            --kmer_database database_output/wildcard/61_kmer_db.json \\
+            --kmer_size 61
+        """
+        copy_db_cmd = "cp $database/card* database_output database_output;cp -r $database/wildcard/ database_output"
+    } else if (!database) { // Download CARD DB if no path given
+        db_download_cmd = "wget https://card.mcmaster.ca/latest/data;tar -xvf data ./card.json -C database_output"
+        db_card_cmd = "rgi card_annotation -i database_output/card.json"
+        database = "."
+        copy_db_cmd = "cp $database/card* database_output"
+    }
```

Comment:
You must not be downloading this within the module, it should be passed into the module via the user! 

This would put a lot of overhead on the nextflow head job, and you could get halfwary through the pipeline for the download to fail and kill the whole thing.There is a lot of hardocded file names which are not safe if they decide to change them!

It would probably make more sense that you pass a database to the wildcard input channel, and a pipeline user must specify the additional flags with `--foo <database_dir_name>/<file_name>`, saying the dir name should be relative to the name of the dir given as inoput or something

Reply:
Okay! I will rewrite. Thanks for the quick suggestions. Btw. locally finally all tests passed for the CARD DB, I was confident to see green tickmarks here as well - of course not the case :sweat_smile:

Reply:
All overhead is now removed. As you suggested, databases can only be provided as input channels. Mandatory: CARD, optional: WildCARD.
nf-test: Since the CARD database is very small (< 4MB) it is downloaded on the fly (by `untar`), it gets processed by the new `rgi/cardannotation` module, and is then ready as input for `rgi/main`. If find this now a really nice and modular flow.
The optional WildCARD database is bigger, so I tested it only locally.